pkg/reconcilermanager/controllers/rsync_status_watch_controller.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/declared"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+)
+
+const (
+	logFieldController    = "controller"
+	logFieldSyncName      = "sync.name"
+	logFieldSyncNamespace = "sync.namespace"
+	logFieldSyncStatus    = "status"
+	logFieldCommit        = "commit"
+
+	syncStatusSucceeded = "SyncSucceeded"
+	syncStatusFailed    = "SyncFailed"
+
+	syncSucceededMessage = "Sync succeeded"
+)
+
+// RsyncStatusWatchController watches the status of RepoSync and RootSync resources.
+type RsyncStatusWatchController struct {
+	client        client.Client
+	logger        *LoggingController
+	statusTracker *StatusTracker
+	syncID        core.ID
+}
+
+// NewRsyncStatusWatchController initializes a new RsyncStatusWatchController.
+func NewRsyncStatusWatchController(mgr manager.Manager, log logr.Logger, id core.ID) *RsyncStatusWatchController {
+	return &RsyncStatusWatchController{
+		client:        mgr.GetClient(),
+		logger:        NewLoggingController(log),
+		statusTracker: NewStatusTracker(),
+		syncID:        id,
+	}
+}
+
+// newKeySelectorPredicate creates a predicate to filter events based on the object key.
+func newKeySelectorPredicate(key client.ObjectKey) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		return client.ObjectKeyFromObject(obj) == key
+	})
+}
+
+// Register the RsyncStatusWatchController with the ReconcilerManager.
+func (c *RsyncStatusWatchController) Register(mgr controllerruntime.Manager) error {
+	cm := controllerruntime.NewControllerManagedBy(mgr).
+		Named("RsyncStatusWatchController").
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		})
+
+	switch c.syncID.Kind {
```

Comment:
Since this controller runs in the reconciler, it only needs to watch one of these types. And it only needs to trigger events for the specific RSync that corresponds to the name & namespace.

If it were external, then it would watch them all, which would be more efficient, but we don't want to run a separate Deployment today.

Reply:
Nevermind. We agreed to have a single controller in the reconciler-manager.